// @sphereon/pe-models/model/filter.ts
/**
 * DIF PE API
 * Decentralized Identity Foundation Presentation Exchange API
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


/**
 * This is the filter that can be applied on the field\'s value to know if the available credential with a specifici field is acceptable or not.
 */
export interface Filter { 
    /**
     * Object type of the acceptable field.
     */
    type: string;
    /**
     * format in which the item is expected. e.g. date time format of some sort.
     */
    format?: string;
    /**
     * The regular expression that can help filter the target object.
     */
    pattern?: string;
    /**
     * acceptable minimum length (e.g. of string) that is acceptable
     */
    minLength?: number;
    /**
     * acceptable maximum length (e.g. of string) that is acceptable
     */
    maxLength?: number;
    /**
     * acceptable minimum value (e.g. 0) that is acceptable for the number including this number (i.e. 0 in this case)
     */
    minimum?: number;
    /**
     * acceptable minimum value (e.g. 0) that is acceptable for the number exclusive of the number (i.e. 0 in this case)
     */
    exclusiveMinimum?: number;
    /**
     * acceptable maximum value (e.g. 100) that is acceptable for the number exclusive of the number (i.e. 100 in this case)
     */
    exclusiveMaximum?: number;
    /**
     * acceptable maximum value (e.g. 100) that is acceptable for the number including this number (i.e. 100 in this case)
     */
    maximum?: number;
    /**
     * A fixed value that is accpetable in this variable
     */
    _const?: number;
    /**
     * Fixed listed values which are acceptable in this variable.
     */
    _enum?: Array<number>;
    /**
     * The values which will make the item unacceptable
     */
    not?: object;
}
