// @sphereon/pe-models/model/inputDescriptor.ts
/**
 * DIF PE API
 * Decentralized Identity Foundation Presentation Exchange API
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { Constraints } from './constraints';
import { Schema } from './schema';


/**
 * Input Descriptor is used by a Verifier to describe the information required of a Holder before an interaction can proceed.
 */
export interface InputDescriptor { 
    /**
     * The verifiable credentials that is acceptable from the holder. e.g. wa_driver_license
     */
    id: string;
    /**
     * The verifiable credentials that is acceptable from the holder in a human readibly string form. i.e. an alternative for the humans to the id
     */
    name?: string;
    /**
     * It describes the purpose for which the Presentation Definition\'s inputs are being requested.
     */
    purpose?: string;
    /**
     * A group from which the specific credential is required.
     */
    group?: Array<string>;
    /**
     * List of schemas for the credential to be compatible with.
     */
    schema: Array<Schema>;
    constraints?: Constraints;
}
