// @sphereon/pe-models/model/presentationDefinition.ts
/**
 * DIF PE API
 * Decentralized Identity Foundation Presentation Exchange API
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { InputDescriptor } from './inputDescriptor';
import { SubmissionRequirement } from './submissionRequirement';
import { Format } from './format';


/**
 * Presentation Definitions are objects that articulate what proofs a Verifier requires.
 */
export interface PresentationDefinition { 
    /**
     * A UUID or some other unique ID to identify this Presentation Definition
     */
    id: string;
    /**
     * A name property is a human-friendly string intended to constitute a distinctive designation of the Presentation Definition.
     */
    name?: string;
    /**
     * It describes the purpose for which the Presentation Definition\'s inputs are being requested.
     */
    purpose?: string;
    format?: Format;
    /**
     * List of requirements for described inputs in input descriptors.
     */
    submission_requirements?: Array<SubmissionRequirement>;
    /**
     * List of descriptions of the required inputs.
     */
    input_descriptors: Array<InputDescriptor>;
}
