// @sphereon/pe-models/model/submissionRequirement.ts
/**
 * DIF PE API
 * Decentralized Identity Foundation Presentation Exchange API
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import { Rules } from './rules';


/**
 * It defines what combinations of inputs a processing entity must submit to comply with the requirements of a Verifier.
 */
export interface SubmissionRequirement { 
    /**
     * what verifiable credential is demanded in order for the presentation to be sensible.
     */
    name?: string;
    /**
     * Why this verifiable credentials is requested.
     */
    purpose?: string;
    rule: Rules;
    /**
     * How many of the credentials are demanded for the presentation to be sensible.
     */
    count?: number;
    /**
     * Minimum of the credentials are demanded for the presentation to be sensible.
     */
    min?: number;
    /**
     * Maximum of the credentials are demanded for the presentation to be sensible.
     */
    max?: number;
    /**
     * One of the group strings specified for one or more [[ref:Input Descriptor Objects].
     */
    from?: string;
    /**
     * nested submission_requirement and is a way to organize the requirements. For example requirements mentioned in the flat list will also be able to provide same information but will be less organized.
     */
    from_nested?: Array<SubmissionRequirement>;
}
