// @sphereon/wellknown-dids-client/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromiseStatusEnum = exports.ServiceTypesEnum = exports.ValidationStatusEnum = exports.ProofFormatTypesEnum = void 0;
var ProofFormatTypesEnum;
(function (ProofFormatTypesEnum) {
    ProofFormatTypesEnum["JSON_WEB_TOKEN"] = "jwt";
    ProofFormatTypesEnum["JSON_LD"] = "jsonld";
})(ProofFormatTypesEnum = exports.ProofFormatTypesEnum || (exports.ProofFormatTypesEnum = {}));
var ValidationStatusEnum;
(function (ValidationStatusEnum) {
    ValidationStatusEnum["VALID"] = "valid";
    ValidationStatusEnum["INVALID"] = "invalid";
})(ValidationStatusEnum = exports.ValidationStatusEnum || (exports.ValidationStatusEnum = {}));
var ServiceTypesEnum;
(function (ServiceTypesEnum) {
    ServiceTypesEnum["LINKED_DOMAINS"] = "LinkedDomains";
})(ServiceTypesEnum = exports.ServiceTypesEnum || (exports.ServiceTypesEnum = {}));
var PromiseStatusEnum;
(function (PromiseStatusEnum) {
    PromiseStatusEnum["REJECTED"] = "rejected";
    PromiseStatusEnum["FULFILLED"] = "fulfilled";
})(PromiseStatusEnum = exports.PromiseStatusEnum || (exports.PromiseStatusEnum = {}));
