// @sphereon/wellknown-dids-client/dist/verifier/WellKnownDidVerifier.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WellKnownDidVerifier = void 0;
const ssi_types_1 = require("@sphereon/ssi-types");
const constants_1 = require("../constants");
const types_1 = require("../types");
const utils_1 = require("../utils");
class WellKnownDidVerifier {
    /** Verifier constructor */
    constructor(config) {
        this.config = config;
    }
    /**
     * Verifies the domain linkage from a DID document.
     *
     * @param args The arguments for verifying domain linkage.
     * @return {IDomainLinkageValidation}, The validation result.
     */
    verifyDomainLinkage(args) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!args.verifySignatureCallback && (!this.config || !((_a = this.config) === null || _a === void 0 ? void 0 : _a.verifySignatureCallback))) {
                return Promise.reject(Error(constants_1.WDCErrors.MUST_SUPPLY_VERIFY_SIGNATURE_CALLBACK));
            }
            // DID document should have a service property
            if (!args.didDocument.service)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SERVICE_NOT_PRESENT });
            // Service property should contain 'LinkedDomains' types
            const linkedDomainsEndpointDescriptors = args.didDocument.service.filter((service) => service.type = types_1.ServiceTypesEnum.LINKED_DOMAINS);
            if (linkedDomainsEndpointDescriptors.length === 0)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SERVICE_NOT_CONTAIN_ANY_SERVICE_WITH_TYPE + `${types_1.ServiceTypesEnum.LINKED_DOMAINS}` });
            const descriptorValidations = linkedDomainsEndpointDescriptors.map((descriptor) => {
                if (descriptor.id.startsWith('#')) {
                    descriptor.id = `${args.didDocument.id}${descriptor.id}`;
                }
                return this.verifyEndpointDescriptor({
                    descriptor,
                    verifySignatureCallback: args.verifySignatureCallback,
                    onlyVerifyServiceDid: args.onlyVerifyServiceDid
                });
            });
            return yield Promise.allSettled(descriptorValidations)
                .then((results) => {
                return {
                    status: results.find((result) => result.status === types_1.PromiseStatusEnum.REJECTED || result.value.status === types_1.ValidationStatusEnum.INVALID) ? types_1.ValidationStatusEnum.INVALID : types_1.ValidationStatusEnum.VALID,
                    endpointDescriptors: results.map((result) => result.status === types_1.PromiseStatusEnum.FULFILLED ? result.value : result.reason)
                };
            });
        });
    }
    /**
     * Verifies the endpoint descriptor.
     *
     * @param args The arguments to verify the descriptor.
     * @return {IDescriptorValidation}, The validation result.
     */
    verifyEndpointDescriptor(args) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!args.verifySignatureCallback && (!this.config || !((_a = this.config) === null || _a === void 0 ? void 0 : _a.verifySignatureCallback))) {
                return Promise.reject(Error(constants_1.WDCErrors.MUST_SUPPLY_VERIFY_SIGNATURE_CALLBACK));
            }
            return this.verifyEndpointDescriptorStructure(args.descriptor).then(() => __awaiter(this, void 0, void 0, function* () {
                const resourceValidations = this.getOrigins(args.descriptor)
                    .map((origin) => (0, utils_1.fetchWellKnownDidConfiguration)(origin)
                    .then((didConfigurationResource) => {
                    var _a;
                    return this.verifyResource({
                        configuration: didConfigurationResource,
                        did: (((_a = this.config) === null || _a === void 0 ? void 0 : _a.onlyVerifyServiceDid) || args.onlyVerifyServiceDid)
                            ? args.descriptor.id
                            : undefined, verifySignatureCallback: args.verifySignatureCallback
                    });
                }));
                return yield Promise.allSettled(resourceValidations)
                    .then((results) => {
                    return {
                        status: results.find((result) => result.status === types_1.PromiseStatusEnum.REJECTED || result.value.status === types_1.ValidationStatusEnum.INVALID) ? types_1.ValidationStatusEnum.INVALID : types_1.ValidationStatusEnum.VALID,
                        resources: results.map((result) => result.status === types_1.PromiseStatusEnum.FULFILLED ? result.value : result.reason)
                    };
                });
            }));
        });
    }
    /**
     * Verifies the DID configuration resource.
     *
     * @param args The arguments to verify the resource.
     * @return {IResourceValidation}, The validation result.
     */
    verifyResource(args) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!args.verifySignatureCallback && (!this.config || !((_a = this.config) === null || _a === void 0 ? void 0 : _a.verifySignatureCallback))) {
                return Promise.reject(Error(constants_1.WDCErrors.MUST_SUPPLY_VERIFY_SIGNATURE_CALLBACK));
            }
            if (args.configuration && args.origin) {
                return Promise.reject(Error(constants_1.WDCErrors.CANT_SUPPLY_BOTH_CONFIGURATION_AND_ORIGIN));
            }
            if (!args.configuration && !args.origin) {
                return Promise.reject(Error(constants_1.WDCErrors.NO_DID_CONFIGURATION_RESOURCE_OR_ORIGIN_SUPPLIED));
            }
            let did;
            if (args.did) {
                did = (0, ssi_types_1.parseDid)(args.did).did;
            }
            if (args.origin) {
                if (new URL(args.origin).protocol !== 'https:')
                    return Promise.reject(constants_1.WDCErrors.ORIGIN_NOT_SECURE);
            }
            const didConfigurationResource = args.configuration
                // @ts-ignore: We know for sure the config is present
                // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
                ? yield (0, utils_1.verifyResourceStructure)(args.configuration).then(() => args.configuration)
                // @ts-ignore: We know for sure the origin is present
                // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
                : yield (0, utils_1.fetchWellKnownDidConfiguration)(args.origin);
            const credentialValidations = didConfigurationResource.linked_dids
                .filter((item) => {
                if (!did)
                    return true;
                let credential;
                if (typeof item === 'string') {
                    try {
                        credential = (0, utils_1.decodeToken)(item, false).vc;
                    }
                    catch (error) {
                        return true;
                    }
                }
                else {
                    credential = item;
                }
                return credential.credentialSubject.id === did;
            })
                .map((credential) => this.verifyDomainLinkageCredential({ credential, verifySignatureCallback: args.verifySignatureCallback }));
            if (credentialValidations.length === 0)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.NO_CREDENTIALS_FOUND_FOR_DID + `${args.did}` });
            return Promise.allSettled(credentialValidations)
                .then((results) => {
                return {
                    status: results.find((result) => result.status === types_1.PromiseStatusEnum.REJECTED) ? types_1.ValidationStatusEnum.INVALID : types_1.ValidationStatusEnum.VALID,
                    credentials: results.map((result) => result.status === types_1.PromiseStatusEnum.FULFILLED ? result.value : result.reason)
                };
            });
        });
    }
    /**
     * Verifies the domain linkage credential.
     *
     * @param args The domain linkage credential. Types can be JWT or JSONLD.
     * @return {ICredentialValidation}, The validation result.
     */
    verifyDomainLinkageCredential(args) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!args.verifySignatureCallback && (!this.config || !((_a = this.config) === null || _a === void 0 ? void 0 : _a.verifySignatureCallback))) {
                return Promise.reject(Error(constants_1.WDCErrors.MUST_SUPPLY_VERIFY_SIGNATURE_CALLBACK));
            }
            if (typeof args.credential === 'string') {
                return this.verifyJsonWebTokenProofFormat(args.credential)
                    .then(() => this.verifyDomainLinkageCredentialStructure((0, utils_1.decodeToken)(args.credential, false).vc))
                    .then(() => args.verifySignatureCallback
                    ? args.verifySignatureCallback({ credential: args.credential, proofFormat: types_1.ProofFormatTypesEnum.JSON_WEB_TOKEN })
                    // @ts-ignore: We know for sure the config is present
                    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
                    : this.config.verifySignatureCallback({ credential: args.credential, proofFormat: types_1.ProofFormatTypesEnum.JSON_WEB_TOKEN }))
                    .then((verificationResult) => {
                    if (!verificationResult.verified)
                        return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.SIGNATURE_IS_INVALID });
                    return { status: types_1.ValidationStatusEnum.VALID };
                });
            }
            return this.verifyDomainLinkageCredentialStructure(args.credential)
                .then(() => args.verifySignatureCallback
                ? args.verifySignatureCallback({ credential: args.credential, proofFormat: types_1.ProofFormatTypesEnum.JSON_LD })
                // @ts-ignore: We know for sure the config is present
                // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
                : this.config.verifySignatureCallback({ credential: args.credential, proofFormat: types_1.ProofFormatTypesEnum.JSON_LD }))
                .then((verificationResult) => {
                if (!verificationResult.verified)
                    return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.SIGNATURE_IS_INVALID });
                return { status: types_1.ValidationStatusEnum.VALID };
            });
        });
    }
    /**
     * Verifies the endpoint descriptor object structure.
     *
     * @param descriptor The endpoint descriptor.
     */
    verifyEndpointDescriptorStructure(descriptor) {
        return __awaiter(this, void 0, void 0, function* () {
            // The object MUST contain an id property
            if (!descriptor.id)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ID_NOT_PRESENT_IN_SERVICE });
            // The object id property value MUST be a valid DID URL reference
            try {
                (0, ssi_types_1.parseDid)(descriptor.id);
            }
            catch (error) {
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ID_NOT_VALID_DID_URL });
            }
            // The object MUST contain a type property
            if (!descriptor.type)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_TYPE_NOT_PRESENT_IN_SERVICE });
            // The object type property value MUST be the string "LinkedDomains".
            if (descriptor.type !== types_1.ServiceTypesEnum.LINKED_DOMAINS)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_TYPE_NOT_CONTAIN_VALID_LINKED_DOMAIN });
            // The object MUST contain a serviceEndpoint property
            if (!descriptor.serviceEndpoint)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SERVICE_ENDPOINT_NOT_PRESENT_IN_SERVICE });
            if (typeof descriptor.serviceEndpoint === 'string') {
                // The object serviceEndpoint property can be a string and the value MUST be an origin string
                if (new URL(descriptor.serviceEndpoint).origin !== descriptor.serviceEndpoint)
                    return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SERVICE_ENDPOINT_NOT_CONTAIN_VALID_ORIGIN });
                if (new URL(descriptor.serviceEndpoint).protocol !== 'https:')
                    return Promise.reject({
                        status: types_1.ValidationStatusEnum.INVALID,
                        message: constants_1.WDCErrors.PROPERTY_ORIGIN_NOT_SECURE
                    });
            }
            if (typeof descriptor.serviceEndpoint === 'object') {
                // The object serviceEndpoint property can be an object which MUST contain an origins property
                if (!Object.prototype.hasOwnProperty.call(descriptor.serviceEndpoint, 'origins'))
                    return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SERVICE_ENDPOINT_NOT_CONTAIN_ORIGIN });
                // The object serviceEndpoint property should have origins
                if (descriptor.serviceEndpoint.origins.length === 0)
                    return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ORIGIN_NOT_CONTAIN_ANY_ORIGIN });
                // The origins should be valid
                for (const origin of descriptor.serviceEndpoint.origins) {
                    if (new URL(origin).origin !== origin)
                        return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ORIGIN_CONTAIN_INVALID_ORIGIN });
                    if (new URL(origin).protocol !== 'https:')
                        return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ORIGIN_NOT_SECURE });
                }
            }
        });
    }
    /**
     * Verifies the structure of a JWT domain linkage credential.
     *
     * @param token The JWT token.
     */
    verifyJsonWebTokenProofFormat(token) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.verifyJsonWebTokenProofHeaderStructure((0, utils_1.decodeToken)(token, true));
            yield this.verifyJsonWebTokenProofPayloadStructure((0, utils_1.decodeToken)(token, false));
        });
    }
    /**
     * Verifies the structure of a JWT domain linkage credential header.
     *
     * @param header The JWT header.
     */
    verifyJsonWebTokenProofHeaderStructure(header) {
        return __awaiter(this, void 0, void 0, function* () {
            // Property kid MUST be present in the JWT Header
            if (!header.kid)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_KID_NOT_PRESENT_IN_JWT_HEADER });
            // Property alg MUST be present in the JWT Header
            if (!header.alg)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ALG_NOT_PRESENT_IN_JWT_HEADER });
            // Additional members MUST NOT be present in the header
            if (Object.getOwnPropertyNames(header).filter(property => !['kid', 'alg'].includes(property)).length > 0)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.JWT_HEADER_CONTAINS_ADDITIONAL_PROPS });
        });
    }
    /**
     * Verifies the structure of a JWT domain linkage credential payload.
     *
     * @param payload The JWT payload.
     */
    verifyJsonWebTokenProofPayloadStructure(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            // Property iss MUST be present in the JWT Payload
            if (!payload.iss)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ISS_NOT_PRESENT_IN_JWT_PAYLOAD });
            // Property sub MUST be present in the JWT Payload
            if (!payload.sub)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SUB_NOT_PRESENT_IN_JWT_PAYLOAD });
            // Property vc MUST be present in the JWT Payload
            if (!payload.vc)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_VC_NOT_PRESENT_IN_JWT_PAYLOAD });
            // Property iss MUST be equal to credentialSubject.id.
            if (payload.vc.credentialSubject && payload.vc.credentialSubject.id !== payload.iss)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ISS_NOT_MATCH_CREDENTIAL_SUBJECT_ID_IN_JWT_PAYLOAD });
            // Property sub MUST be equal to credentialSubject.id.
            if (payload.vc.credentialSubject && payload.vc.credentialSubject.id !== payload.sub)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_SUB_NOT_MATCH_CREDENTIAL_SUBJECT_ID_IN_JWT_PAYLOAD });
            // Additional members MUST NOT be present in the payload
            if (Object.getOwnPropertyNames(payload).filter(property => !['exp', 'iss', 'nbf', 'sub', 'vc'].includes(property)).length > 0)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.JWT_PAYLOAD_CONTAINS_ADDITIONAL_PROPS });
        });
    }
    /**
     * Verifies the structure of a domain linkage credential.
     *
     * @param credential The domain linkage credential.
     */
    verifyDomainLinkageCredentialStructure(credential) {
        return __awaiter(this, void 0, void 0, function* () {
            // Property issuanceDate MUST be present.
            if (!credential.issuanceDate)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ISSUANCE_DATE_NOT_PRESENT_IN_CREDENTIAL });
            // Property issuanceDate MUST be a valid date.
            if ( /*typeof credential.issuanceDate === 'string' && */isNaN(Date.parse(credential.issuanceDate)))
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_ISSUANCE_DATE_NOT_VALID });
            // Property expirationDate MUST be present.
            if (!credential.expirationDate)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_EXPIRATION_DATE_NOT_PRESENT_IN_CREDENTIAL });
            // Property expirationDate MUST be a valid date.
            if ( /*typeof credential.expirationDate === 'string' && */isNaN(Date.parse(credential.expirationDate)))
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_EXPIRATION_DATE_NOT_VALID });
            // Property credentialSubject MUST be present.
            if (!credential.credentialSubject)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_NOT_PRESENT_IN_CREDENTIAL });
            // Property credentialSubject.id MUST be present.
            if (!credential.credentialSubject.id)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ID_NOT_PRESENT_IN_CREDENTIAL });
            // Property credentialSubject.id MUST be a DID.
            try {
                (0, ssi_types_1.parseDid)(credential.credentialSubject.id);
            }
            catch (error) {
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ID_NOT_VALID_DID });
            }
            // The credentialSubject.id value MUST be equal to the issuer of the Domain Linkage Credential.
            if (credential.issuer && credential.credentialSubject.id !== credential.issuer)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ID_NOT_MATCH_ISSUER });
            // Property credentialSubject.origin MUST be present.
            if (!credential.credentialSubject.origin)
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ORIGIN_NOT_PRESENT_IN_CREDENTIAL });
            // Property credentialSubject.origin MUST be a domain Origin.
            try {
                if (new URL(credential.credentialSubject.origin).origin !== credential.credentialSubject.origin)
                    return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ORIGIN_NOT_VALID });
            }
            catch (error) {
                return Promise.reject({ status: types_1.ValidationStatusEnum.INVALID, message: constants_1.WDCErrors.PROPERTY_CREDENTIAL_SUBJECT_ORIGIN_NOT_VALID });
            }
        });
    }
    /**
     * Retrieves the origins of an endpoint descriptor.
     *
     * @param descriptor The endpoint descriptor.
     */
    getOrigins(descriptor) {
        if (typeof descriptor.serviceEndpoint === 'string') {
            return [descriptor.serviceEndpoint];
        }
        else {
            // This break with did-resolver@3
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return descriptor.serviceEndpoint.origins;
        }
    }
}
exports.WellKnownDidVerifier = WellKnownDidVerifier;
